Parser/asdl_c.py

```diff
@@ -1150,8 +1150,194 @@ def visitModule(self, mod):
     {NULL}
 };
 
+static PyObject *
+ast_repr_max_depth(AST_object *self, int depth);
+
+static PyObject *
+ast_repr_list(PyObject *list, int depth)
+{
+    assert(PyList_Check(list) || PyTuple_Check(list));
+
+    struct ast_state *state = get_ast_state();
+    if (state == NULL) {
+        return NULL;
+    }
+
+    Py_ssize_t length = PySequence_Size(list);
+    if (length < 0) {
+        return NULL;
+    } else if (length == 0) {
+        return PyObject_Repr(list);
+    }
+
+    _PyUnicodeWriter writer;
```

Comment:
Any reason not to use a pointer here?

Reply:
Wouldn't I then need to dynamically allocate the writer struct?

Reply:
Maybe I should have been more precise: any reason not to use the *public* API which creates the pointer (i.e., `PyUnicodeWriter *writer = PyUnicodeWriter_Create(estimate);` and not `_PyUnicodeWriter`). You can then cast it to `(_PyUnicodeWriter *)(writer)` and use `_PyUnicodeWriter_*` functions if you have guarantees on the objects being passed to the call.

(This API is fairly recent and maybe the previous way to use the unicode writer was as you did).

Reply:
I had no idea that API existed.. (probably because I started working on this before it was added :sweat_smile: )
I switched most things to use the public API, The only method I kept is `_PyUnicodeWriter_WriteASCIIString` because I didn't find an equivalent for it in the public API.

Reply:
I should definitely more precise when I'm doing my reviews... Actually, I have three proposals but I'm not sure which one is preferred:

- Use the public API normally as you did, as if you were not a CPython dev but rather a CPython user. If performances do not really matter here, I think it's fine to keep the checks.
- Use the public API for creating the object for you. Using the public API allows you to have the overallocation feature to avoid resizing the buffer too many times (since you are writing a tree, I think it's good to have that feature automatically managed, instead of configuring the writer yourself). However, use the **private** API for calls that you *know* they don't need checks (like, if you know that you are writing an ASCII string, there is no need for using the UTF8 version because it does additional checks that are not needed I think; but for that you need to cast the writer from `PyUnicodeWriter*` to `_PyUnicodeWriter*`).
- Revert my suggestion and just use the old private API.

@vstinner What do you suggest in general for code in CPython itself. Should we just short-circuit some checks because we *know* that they would not be needed or should we use the public API as if we were an end user?

Reply:
The private C API can be a little bit more efficient. It should only be used in CPython itself. You can run some benchmarks to compare.

Reply:
Ok, I'll run some tests.

Reply:
Ok I have some results:

Private API (with `overallocate` enabled):
```bash
./python -m timeit -s "import ast; tree = ast.parse('def ast_repr_test(a = 1, b = 2, c = 3): return a + b + c')" 'repr(tree)'
20000 loops, best of 5: 18 usec per loop
```
Public API:
```bash
./python -m timeit -s "import ast; tree = ast.parse('def ast_repr_test(a = 1, b = 2, c = 3): return a + b + c')" 'repr(tree)'
10000 loops, best of 5: 20.2 usec per loop
```
The private API appears to be slighly faster. So do you think I should stick with the private API?

Reply:
I think we can keep the public API. Less errors, and this part of the code is not really a critical part I think (generally, `repr()` is for debugging). Out of curiosity, is it a debug, release or PGO build?

Reply:
The results are for the debug build

Reply:
Here are the results for an optimized build (with `./configure --enable-optimizations`):

Private API:
```
100000 loops, best of 5: 3.44 usec per loop
```

Public API:
```
50000 loops, best of 5: 4.16 usec per loop
```

Reply:
I think we can keep the public API. A 0.5us gain for such operations is not worth it IMO. By the way, if you want more precise benchmarks, I can suggest you to use https://github.com/psf/pyperf instead. They give you nice benchmarks and allows you to compare the results, and also includes the stdev which is sometimes of interest.

Reply:
I suggest to use the private API since it's faster.

Reply:
(I like how Victor and me have different suggestions but since he's the one that wrote the PyUnicodeWriter API, I'd suggest following his advice then)

Reply:
@picnixz Your review is appreciated regardless! :) I switched back to the private API :sweat_smile:

Reply:
I'm sorry for this back and forth :') at least, now you know that you have different approaches (and that release or PGO builds should be preferred for benchmarks)